diff -Naur absl/synchronization/BUILD.bazel absl/synchronization/BUILD.bazel
--- absl/synchronization/BUILD.bazel
+++ absl/synchronization/BUILD.bazel
@@ -54,6 +54,9 @@
         "internal/waiter.cc",
         "notification.cc",
     ] + select({
+        "@com_google_asylo//asylo": [
+            "internal/mutex_nonprod.cc",
+        ],
         "//conditions:default": ["mutex.cc"],
     }),
     hdrs = [
