diffBUILD.bazel googletest-release-1.8.1/BUILD.bazel
--- BUILD.bazel
+++ BUILD.bazel
@@ -77,6 +77,7 @@
         {
             ":windows": [],
             ":windows_msvc": [],
+            "@com_google_asylo//asylo": [],
             "//conditions:default": ["-pthread"],
         },
     ),
@@ -97,6 +98,7 @@
     linkopts = select({
         ":windows": [],
         ":windows_msvc": [],
+        "@com_google_asylo//asylo": [],
         "//conditions:default": [
             "-pthread",
         ],
